crypto-mcl.ts

```diff
@@ -0,0 +1,188 @@
+        return {ent: mpk, pEnt: nonce, mtr};
+    }
+
+    /**
+     * Creates a private record for the user to store in her phone.
+     * @param entry as presented in the QRentry
+     * @param proofEntry as presented in the QRentry
+     * @param info as presented in the QRentry
+     * @param counter as hours since the unix epoch
+     * @param aux free string
+     * @return user record to be stored
+     */
+    static scan(entry: mcl.G2, proofEntry: Uint8Array, info: string, counter: number, aux: string): IUserRecord{
+        const r1 = new mcl.Fr();
+        r1.setByCSPRNG();
+        const r2 = new mcl.Fr();
+        r2.setByCSPRNG();
+        const k = crypto_secretbox_keygen();
+        const nonce = secretbox_nonce();
+        const c = crypto_secretbox_easy(from_string(aux), nonce, k);
+
+        const R1 = mcl.mul(mcl.baseG2(), r1);
+        const H1inc = this.infoNonceCounter(info, proofEntry, counter);
+        const Z1 = mcl.pairing(mcl.mul(H1inc, r1), entry);
+        const R2 = mcl.mul(mcl.baseG2(), r2);
+        const Z2 = mcl.pairing(mcl.mul(H1inc, r2), entry);
+        const K = mcl.xor(k, mcl.hashT(Z2));
+
+        return {R1, Z1, R2, K, c, nonce};
+    }
+
+    /**
+     * Convenience method to calculate the hash-to-G1 of the info, nonce, and counter.
+     * @param info
+     * @param nonce
+     * @param counter
+     */
+    static infoNonceCounter(info: string, nonce: Uint8Array, counter: number): mcl.G1 {
+        const counterStr = counter.toString();
+        const infoNonceCnt = new Uint8Array(info.length + nonce.length + counterStr.length);
+        infoNonceCnt.set(from_string(info));
+        infoNonceCnt.set(nonce, info.length);
+        infoNonceCnt.set(from_string(counterStr), info.length + nonce.length);
+        return mcl.hashAndMapToG1(infoNonceCnt);
+    }
+
+    /**
+     * Generates a pre-trace by the location owner, one for every slot of possible infection.
+     * @param mtr master trace record from the location
+     * @param cnt hours since the unix epoch
+     * @return an IPreTrace for the health authority
+     */
+    static genPreTrace(mtr: IMasterTrace, cnt: number): IPreTrace{
+        const {info, nonce, mskv, mskHAEnc: ctxt} = mtr;
+        const Tprime = mcl.mul(this.infoNonceCounter(info, nonce, cnt), mskv);
+        return {Tprime, info, nonce, ctxt};
+    }
+
+    /**
+     * Generates an anonymous trace that can only be used by people having scanned the QRentry code
+     * @param skH the health authorities private key
+     * @param cnt hours since the unix epoch
+     * @param ptr from genPreTrace
+     * @param proofPTR is ignored for the moment
+     * @return tracing information to be stored
+     */
+    static genTrace(skH: Uint8Array, cnt: number, ptr: IPreTrace, proofPTR?: Uint8Array): mcl.G1{
+        const {Tprime, info, nonce, ctxt} = ptr;
+        const mskHA = new mcl.Fr();
+        const pkH = crypto_scalarmult_base(skH)
```

Comment:
I this case, I would just store key_pair for HA, so that is the only thing you need to pass in, that "saves" this operation, but more importantly, makes the code a little bit clearer

Reply:
done.